### PR TITLE
PHP 8 support

### LINKED ISSUE #44: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by ocean**
Updating to add PHP 8.0 support for #44 

First draft of PR for TravisCI testing.

### CODE REVIEW

**Review by boesing — CHANGES_REQUESTED** [review]
LGTM so far.

You might want to have a look on the failing unit tests.
Some methods got dropped in phpunit 9 so we have to find a better way to verify what've happened.

**Comment by ocean** [comment]
@boesing thanks. Yep, I'm having a look at the breaking tests on Travis at the moment.

There are deprecations in PHP-CodeSniffer too, which is brought in by `laminas/laminas-coding-standard` which I noticed was at version `~1.0.0` still - I've just updated that in Composer to version `~2.1.0` and will try it out now - does that sound ok?

**Comment by boesing** [comment]
I would prefer not upgrading codestyle in here to have a clear overview on what've changed or PHP 8.0.

If there are issues with codesniffer, we can bump dev-dependency for that package to ensure it always works with PHP 8.0.
But afaik, codestyle should not be checked on 8.0 so there is probably no issue but locally?

**Comment by boesing** [comment]
@ocean If I can help you, just ping me 👍

**Comment by ocean** [comment]
Sorry I haven't had a chance to get back to this until now @boesing 

I would happy with some help. The test issue I've found most annoying is the lack of ability to access protected properties e.g. here https://travis-ci.com/github/laminas/laminas-http/jobs/400060420#L555

**Comment by boesing** [comment]
> Sorry I haven't had a chance to get back to this until now @boesing
> 
> I would happy with some help. The test issue I've found most annoying is the lack of ability to access protected properties e.g. here https://travis-ci.com/github/laminas/laminas-http/jobs/400060420#L555

No problem, we all have plenty of stuff to do 💪 
You could test the `auth` property by writing a little private "helper" method to extract the value of a property which is not accessible. However, we might want to keep in mind that testing non-public properties is not how tests should work.

So, you could do sth. like this:

```php
$valueFromProperty = $this->extractPropertyValue($objectYouWantToExtractThePropertyFrom, 'propertyName');
self::assertEquals($expectedValue, $valueFromProperty);
```

Within the `extractPropertyValue` "helper" method, you can use [ReflectionProperty](https://www.php.net/reflectionproperty) to extract the value. Have a look at the `ReflectionProperty::setAccessible` method.

**Comment by ocean** [comment]
@boesing GitHub still says there's a review to do on this PR, but I can't see one.

Also it seems the TravisCI jobs for this repo have stopped running 😞  I ran tests on my local machine with PHP 7.4 and they seemed ok, but am keen to see what Travis does with PHP latest.

**Comment by boesing** [comment]
@ocean Thats fine, we have to review manually with each supported PHP version as we ran out of build hours on travis.
Sorry for the delay tho.

**Review by boesing — APPROVED** [review]
LGTM.
Have to manually check locally with each PHP version. Wont make it this week tho.

**Comment by ocean** [comment]
@boesing I've got Travis builds running in my branch now, and the only thing failing is the Curl "test chunked response performance" test in PHP 8:

https://travis-ci.com/github/ocean/laminas-http/jobs/452211509#L709

I'm thinking there may be some issue that could be fixed by an upgrade/change in the curl library, but haven't had a look yet.

File: .travis.yml [hunk 1]
```diff
@@ -16,41 +16,25 @@ env:
...
-      env:
-        - DEPS=lowest
-    - php: 7.1
+    - php: 7.4
       env:
         - DEPS=latest
         - CS_CHECK=true
         - TEST_COVERAGE=true
-    - php: 7.2
+    - php: nightly
```
Comment (boesing) [inline .travis.yml:40]: Could you change the travis configuration as in laminas/laminas-servicemanager#64?
Travis added PHP 8.0 [yesterday](https://travis-ci.community/t/php-8-0-missing/10132/23) support for the matrix and thus we can add `8.0` aswell as `nightly` where nightly is allowed to fail.

There is an example in the linked PR 👍

**Review by boesing — CHANGES_REQUESTED** [review]
> https://travis-ci.com/github/ocean/laminas-http/jobs/452211509#L709

Hm, thats a tough one as I am not that deep into that stuff and TBH I don't have time to dig into that. 
Maybe @domoran could provide some insights here as he added this test 4 years ago?

File: .travis.yml (hunk 1 above)
Comment (ocean) [reply .travis.yml:40]: Cool, I've added the PHP 8 specific tests in Travis, and allowed `nightly` to fail.

I've had a bit of a look at the PHP Curl library and can't find much info about what constitutes good performance. An easy way to fix the test would be to change the timeout that it's tested against from 20 to 25 or 30, hmm.

**Comment by ocean** [comment]
@boesing I updated the curl chunked response test to test against "less than 25 milliseconds completion" instead of 20 milliseconds. This has made the test pass, meaning we can move on from this hopefully.

**Comment by samsonasik** [comment]
@ocean I tried in my local dev with php 8 and it working ok 👍 

Screen Shot 2020-12-21 at 5 31 25 PM

**Review by samsonasik — APPROVED** [review]

**Comment by domoran** [comment]
Please note, that the 20, 25 in this test do not mean "ms" but it defines the factor on how long it takes to parse the chunked response, compared to the baseline [2000 chunks a 1 byte]   vs    [2000 chunks a 1 byte + 1 chunk 10 MB]. Its kind of hard to make a prediction on how long this is supposed to take - in the original implementation due to a very bad string handling the 10 MB chunk was copied thousands of times, making the packet take multiple minutes to parse. 

Since a test was required to perform the change on the zend http library I made this test to reproduce the problem and see that it was fixed - however the test has only the purpose to make sure the specific problem of the old implementation is gone.  So there is nothing wrong with increasing the asserted ratio as long as it is ensured that the absolute times of parsing such a response are still reasonable (maybe sub 500ms?). 

Please also note, that the response should have never been loaded into memory anyways, because with large responses this will break any memory limit anyway. But that change was too big to perform at that time. So I would suggest switching to another HTTP library anyway, that does not suffer from this issue.

File: phpunit.xml.dist [hunk 2]
```diff
@@ -1,52 +1,42 @@
...
-        <whitelist addUncoveredFilesFromWhitelist="true">
-            <directory suffix=".php">./src</directory>
-        </whitelist>
-    </filter>
-
-    <php>
-        <ini name="date.timezone" value="UTC"/>
-
-        <!-- OB_ENABLED should be enabled for some tests to check if all
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" bootstrap="vendor/autoload.php" colors="true">
```
Comment (froschdesign) [inline phpunit.xml.dist]: The control of the version is done in the Composer configuration file. Therefore the current installed XML schema definition can be used here.

```suggestion
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd" bootstrap="vendor/autoload.php" colors="true">
```

**Comment by ocean** [comment]
@domoran Thanks very much for the clarification. I see, so it is more like "this test should take less than 20 times as long as parsing the initial baseline response".

@samsonasik Thanks for your double-checking input too. The test was passing on my local as well, but failing on Travis (but only by one or two increments), so I guess it's mostly caused by the limited speed of the Travis CI servers.

@froschdesign Thanks, updated that schema version string now, ready for review 👍

**Review by boesing — APPROVED** [review]

**Comment by boesing** [comment]
Thanks @domoran for spending time on this! Thats really appreciated.
@ocean I'll try to create a new release next week between Christmas and NYE.

**Comment by samsonasik** [comment]
@boesing it is already next week, I will merge it right now ;)

**Comment by samsonasik** [comment]
Oh, the milestone is 2.12, automatic release is not working

**Comment by samsonasik** [comment]
trying create milestone 2.14.0 and re-close the milestone

**Comment by samsonasik** [comment]
2.14.0 released https://github.com/laminas/laminas-http/releases/tag/2.14.0 🎉 
Thank you all ;)

**Comment by SamuelNitsche** [comment]
May I ask why this was released without a PHP 8 compatible version of [laminas/laminas-validator](https://github.com/laminas/laminas-validator)?

**Comment by Ocramius** [comment]
@SamuelNitsche one thing at a time - https://github.com/laminas/laminas-validator/pull/75 just needs reviews, if you got time, and then it can certainly be released IMO

**Comment by SamuelNitsche** [comment]
Got it, thanks @Ocramius!

**Comment by boesing** [comment]
Thanks @samsonasik 😄
and yes, but this week has still 4 days left 😉

**Pull request merged**